examples/module/mpc/cartpole.py

```diff
@@ -0,0 +1,146 @@
+    n_sc = n_state + n_ctrl
+
+    class CartPole(pp.module.NLS):
+        def __init__(self, dt, length, cartmass, polemass, gravity):
+            super().__init__()
+            self.tau = dt
+            self.length = length
+            self.cartmass = cartmass
+            self.polemass = polemass
+            self.gravity = gravity
+            self.polemassLength = self.polemass * self.length
+            self.totalMass = self.cartmass + self.polemass
+
+        def state_transition(self, state, input, t=None):
+            x, xDot, theta, thetaDot = state.squeeze()
+            force = input.squeeze()
+            costheta = torch.cos(theta)
+            sintheta = torch.sin(theta)
+
+            temp = (
+                force + self.polemassLength * thetaDot**2 * sintheta
+            ) / self.totalMass
+            thetaAcc = (self.gravity * sintheta - costheta * temp) / (
+                self.length * (4.0 / 3.0 - self.polemass * costheta**2 / self.totalMass)
+            )
+            xAcc = temp - self.polemassLength * thetaAcc * costheta / self.totalMass
+
+            _dstate = torch.stack((xDot, xAcc, thetaDot, thetaAcc))
+
+            return (state.squeeze() + torch.mul(_dstate, self.tau)).unsqueeze(0)
+
+        def observation(self, state, input, t=None):
+            return state
+
+    expert_seed = 7
+    assert expert_seed != args.seed
+    torch.manual_seed(expert_seed)
+
+    Q = torch.tile(torch.eye(n_state + n_ctrl, device=device), (n_batch, T, 1, 1))
+    p = torch.randn(n_batch, T, n_sc)
+    dt = 0.01
+    g = 9.81
+    time  = torch.arange(0, T, device=device) * dt
+
+    expert = dict(
+        Q = torch.tile(torch.eye(n_state + n_ctrl, device=device), (n_batch, T, 1, 1)),
+        p = torch.randn(n_batch, T, n_sc).to(device),
+        len = torch.tensor(1.5).to(device),
+        m_cart = torch.tensor(20).to(device),
+        m_pole = torch.tensor(10).to(device),
+    )
+    fname = os.path.join(args.save, 'expert.pkl')
+    with open(fname, 'wb') as fi:
+        pkl.dump(expert, fi)
+
+    current_u = torch.sin(time).unsqueeze(1).unsqueeze(0)
+    x_init = torch.tensor([[0, 0, torch.pi, 0]], device=device)
+    expert_cartPoleSolver = CartPole(dt, expert['len'], expert['m_cart'], expert['m_pole'], g).to(device)
+
+    torch.manual_seed(args.seed)
+    len = torch.tensor(1.55).to(device).requires_grad_()
+    m_cart = torch.tensor(20.3).to(device).requires_grad_()
+    m_pole = torch.tensor(10.4).to(device).requires_grad_()
+
+    fname = os.path.join(args.save, 'cartpole losses.csv')
+    loss_f = open(fname, 'w')
+    loss_f.write('im_loss,mse\n')
+    loss_f.flush()
+
+    mpc_expert = pp.module.MPC(expert_cartPoleSolver, T, step=15).to(device)
+    x_true, u_true, cost_true = mpc_expert(expert['Q'], expert['p'], x_init, dt, current_u)
+
+    def get_loss(_len, _m_cart, _m_pole):
+        x_init = torch.tensor([[0, 0, torch.pi, 0]], device=device)
```

Comment:
Why are you using a fixed x_init? There will be no meaning to do 1000 iteration....